sql/src/main/java/io/crate/planner/consumer/ManyTableConsumer.java

```diff
@@ -0,0 +1,337 @@
+
+        @Override
+        public Void visitRelationColumn(RelationColumn relationColumn, Set<QualifiedName> context) {
+            context.add(relationColumn.relationName());
+            return null;
+        }
+    }
+
+    private static class RelationColumnReWriteCtx {
+        private final QualifiedName newName;
+        private final QualifiedName left;
+        private final QualifiedName right;
+        private final int rightOffset;
+
+        public RelationColumnReWriteCtx(TwoTableJoin join) {
+            this(join.name(), join.leftName(), join.rightName(), join.left().querySpec().outputs().size());
+        }
+
+        public RelationColumnReWriteCtx(QualifiedName newName, QualifiedName left, QualifiedName right, int rightOffset) {
+            this.newName = newName;
+            this.left = left;
+            this.right = right;
+            this.rightOffset = rightOffset;
+        }
+    }
+
+    private static class RelationColumnReWriter extends ReplacingSymbolVisitor<RelationColumnReWriteCtx> {
+
+        private static final RelationColumnReWriter INSTANCE = new RelationColumnReWriter(false);
+
+        public RelationColumnReWriter(boolean inPlace) {
+            super(inPlace);
+        }
+
+        @Override
+        public Symbol visitRelationColumn(RelationColumn relationColumn, RelationColumnReWriteCtx context) {
+            if (relationColumn.relationName().equals(context.left)) {
+                return new RelationColumn(context.newName, relationColumn.index(), relationColumn.valueType());
+            }
+            if (relationColumn.relationName().equals(context.right)) {
+                return new RelationColumn(context.newName, relationColumn.index() + context.rightOffset, relationColumn.valueType());
+            }
+            return super.visitRelationColumn(relationColumn, context);
+        }
+    }
+
+    private static class FieldToRelationColumnCtx {
+        private final Map<AnalyzedRelation, QualifiedName> relationToName;
+        private final MultiSourceSelect mss;
+
+        public FieldToRelationColumnCtx(MultiSourceSelect mss) {
+            relationToName = new IdentityHashMap<>(mss.sources().size());
+            for (Map.Entry<QualifiedName, MultiSourceSelect.Source> entry : mss.sources().entrySet()) {
+                relationToName.put(entry.getValue().relation(), entry.getKey());
+            }
+            this.mss = mss;
+        }
+    }
+
+    private static class FieldToRelationColumnVisitor extends ReplacingSymbolVisitor<FieldToRelationColumnCtx> {
+
+        private static final FieldToRelationColumnVisitor INSTANCE = new FieldToRelationColumnVisitor(true);
+
+        public FieldToRelationColumnVisitor(boolean inPlace) {
+            super(inPlace);
+        }
+
+        @Override
+        public Symbol visitField(Field field, FieldToRelationColumnCtx ctx) {
+            QualifiedName qualifiedName = ctx.relationToName.get(field.relation());
+            int idx = 0;
```

Comment:
The RelationColumn needs to point to the place inside the sub querySpec (left or right releation) and not to the place of the MSS querySpec